ui/src/app/machines/views/MachineDetails/MachineConfiguration/PowerForm/PowerForm.test.tsx

```diff
@@ -0,0 +1,142 @@
+        statuses: machineStatusesFactory({
+          abc123: machineStatusFactory(),
+        }),
+      }),
+    });
+  });
+
+  it("is not editable if machine does not have edit permission", () => {
+    state.machine.items[0].permissions = [];
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <PowerForm systemId="abc123" />
+      </Provider>
+    );
+
+    expect(wrapper.find("button[data-test='edit-power-config']").exists()).toBe(
+      false
+    );
+  });
+
+  it("is disabled until the edit button is pressed", () => {
+    state.machine.items[0].permissions = ["edit"];
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <PowerForm systemId="abc123" />
+      </Provider>
+    );
+
+    expect(wrapper.find("FormikForm").prop("editable")).toBe(false);
+
+    wrapper.find("button[data-test='edit-power-config']").simulate("click");
+
+    expect(wrapper.find("FormikForm").prop("editable")).toBe(true);
+  });
+
+  it("initialises the form with machine power parameters", () => {
+    state.general.powerTypes.data = [
+      powerTypeFactory({
+        fields: [
+          powerFieldFactory({ name: "field1" }),
+          powerFieldFactory({ name: "field2" }),
+        ],
+        name: "power-type",
+      }),
+    ];
+    state.machine.items = [
+      machineDetailsFactory({
+        power_parameters: {
+          field1: "value1",
+          field2: "value2",
+        },
+        power_type: "power-type",
+        system_id: "abc123",
+      }),
+    ];
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <PowerForm systemId="abc123" />
+      </Provider>
+    );
+
+    expect(wrapper.find("Formik").prop("initialValues")).toStrictEqual({
+      powerType: "power-type",
+      powerParameters: { field1: "value1", field2: "value2" },
+    });
+  });
+
+  it("correctly dispatches an action to update a machine's power", () => {
+    const machine = machineDetailsFactory({
+      permissions: ["edit"],
+      system_id: "abc123",
+    });
+    state.machine.items = [machine];
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <PowerForm systemId="abc123" />
+      </Provider>
+    );
+
+    wrapper.find("Formik").prop("onSubmit")({
+      powerParameters: {},
+      powerType: "manual",
+    });
+
+    expect(
+      store.getActions().find((action) => action.type === "machine/update")
+    ).toStrictEqual({
+      meta: {
+        method: "update",
+        model: "machine",
+      },
+      payload: {
+        params: {
+          extra_macs: machine.extra_macs,
+          power_parameters: {},
+          power_type: "manual",
+          pxe_mac: machine.pxe_mac,
+          system_id: machine.system_id,
+        },
+      },
+      type: "machine/update",
+    });
+  });
+});
```

Comment:
Might also be good to add a couple of tests covering the QA issues I found (resetting fields etc.) so that we catch any regressions.

Reply:
Good call. I've added tests for:
- Power fields being set to defaults when the power type is changed
- MachineForm/PowerForm only showing errors in editing state
- MachineForm/PowerForm being reset to initial values when for is cancelled
- MachineForm/PowerForm dispatching a machine cleanup action on cancel